Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift

```diff
@@ -0,0 +1,283 @@
+  }
+
+  mutating func finish() -> FinishAction {
+    switch self.state {
+      case .idle:
+        self.state = .terminated(.finished)
+        return .none
+
+      case .waitingForConsumer(let suspendedProducers):
+        self.state = .terminated(.finished)
+        return .resumeProducers(continuations: suspendedProducers.map { $0.continuation })
+
+      case .waitingForProducer(let consumerContinuations):
+        self.state = .terminated(.finished)
+        return .resumeConsumers(continuations: consumerContinuations.map { $0.continuation })
+
+      case .modifying:
+        preconditionFailure("Invalid state.")
+
+      case .terminated:
+        return .none
+    }
+  }
+
+  enum FailAction {
+    case none
+    case resumeProducers(continuations: [UnsafeContinuation<Void, Never>?])
+    case resumeConsumers(continuations: [UnsafeContinuation<Element?, any Error>?])
+  }
+
+  mutating func fail(error: Error) -> FailAction where Failure == Error {
+    switch self.state {
+      case .idle:
+        self.state = .terminated(.failed(error))
+        return .none
+
+      case .waitingForConsumer(let suspendedProducers):
+        self.state = .terminated(.failed(error))
+        return .resumeProducers(continuations: suspendedProducers.map { $0.continuation })
+
+      case .waitingForProducer(let consumerContinuations):
+        self.state = .terminated(.finished)
+        return .resumeConsumers(continuations: consumerContinuations.map { $0.continuation })
+
+      case .modifying:
+        preconditionFailure("Invalid state.")
+
+      case .terminated:
+        return .none
+    }
+  }
+
+  enum ProducerWasCancelledAction {
+    case none
+    case resumeProducer(continuation: UnsafeContinuation<Void, Never>?)
+  }
+
+  mutating func producerWasCancelled(producerId: Int) -> ProducerWasCancelledAction {
+    switch self.state {
+      case .idle:
+        return .none
+
+      case .waitingForConsumer(var suspendedProducers):
+        let placeHolder = SuspendedProducer.placeHolder(id: producerId)
+        self.state = .modifying
+
+        guard let removed = suspendedProducers.remove(placeHolder) else {
+          self.state = .waitingForConsumer(suspendedProducers: suspendedProducers)
+          return .none
+        }
+
+        if suspendedProducers.isEmpty {
+          self.state = .idle
+        } else {
+          self.state = .waitingForConsumer(suspendedProducers: suspendedProducers)
+        }
+        return .resumeProducer(continuation: removed.continuation)
+
+      case .waitingForProducer:
+        return .none
+
+      case .modifying:
+        preconditionFailure("Invalid state.")
+
+      case .terminated:
+        return .none
+    }
+  }
+
+  enum NewConsumerAction {
+    case suspend
+    case resumeConsumerWithElement(Element?)
+    case resumeConsumerWithError(Error)
+    case resumeProducerAndConsumer(UnsafeContinuation<Void, Never>?, Element?)
+  }
+
+  mutating func newConsumer(
```

Comment:
Would prefer if we rename this to `next`

Reply:
I've renames pretty much all the functions, please advise.